src/foam/u2/detail/AbstractSectionedDetailView.js

```diff
@@ -102,6 +106,214 @@ foam.CLASS({
 
         return sections;
       }
+    },
+    {
+      class: 'Boolean',
+      name: 'loading',
+      documentation: `True while we're calculating 'sections'.`,
+      expression: function(outstandingCalculations_) {
+        return outstandingCalculations_ > 0;
+      }
+    },
+    {
+      class: 'FObjectArray',
+      of: 'foam.layout.Section',
+      name: 'sections',
+      factory: function() {
+        return this.sections_;
+      }
+    },
+    {
+      class: 'Long',
+      name: 'outstandingCalculations_',
+      documentation: 'Used alongside `loading` to hide the view while it is loading',
+      factory: function() {
+        return 0;
+      }
+    },
+    {
+      class: 'Long',
+      name: 'nextQueuePosition_',
+      value: 1, // Start at a value greater than 0 so the first one can succeed.
+      documentation: 'Used to avoid race conditions where old results can clobber newer results.'
+    },
+    {
+      class: 'Long',
+      name: 'queuePositionWhenLastUpdated_',
+      documentation: 'Used to avoid race conditions where old results can clobber newer results.'
+    }
+  ],
+
+  methods: [
+    function init() {
+      this.onDetach(this.sections_$.sub(this.updateSections));
+      this.onDetach(this.data$.sub(this.updateSections));
+      this.updateSections();
+    }
+  ],
+
+  listeners: [
+    function updateSections() {
+      /**
+       * This function sets 'sections' to a filtered version of 'sections_'
+       * after checking a few different things. The goal is to do all of the
+       * work to determine which sections should be visible in one place, then
+       * set 'sections' to that set of sections once we know what it is. Then
+       * views that extend AbstractSectionedDetailView can simply use a slot of
+       * 'sections' and don't need to worry about doing all of this work
+       * themselves.
+       * 
+       * The things we need to check are:
+       *   1. That the user has permission to see the section if the section is
+       *      configured that way.
+       *   2. That the section is visible based on the data and the section's
+       *      'isAvailable' method if there is one.
+       *   3. That the section has at least one visible property or action,
+       *      which is based on:
+       *        (a) Whether or not the property or action requires permission to
+       *            be seen.
+       *        (b) Whether or not the action is visible based on the
+       *            'isAvailable' method of the action.
+       *        (c) Whether or not the property is visible based on the
+       *            controllerMode, visibility, or visibilityExpression.
+       */
```

Comment:
Can listeners be FOAM method objects with a `documentation` property?

Reply:
You can put listeners in long-form with documentation:
{
  name: 'updateSections',
  documentation: '...',
  code: function() {...}
}

Reply:
Addressed by 1d8e5c759.